### PR TITLE
[interop] clang name lookup should find declarations in inline namesp…

### PR DESCRIPTION
**Pull request by hyp**
…aces

### CODE REVIEW

**Comment by hyp** [comment]
@swift-ci please test

File: lib/ClangImporter/ClangImporter.cpp
Comment (zoecarver) [inline lib/ClangImporter/ClangImporter.cpp:4278]: If this API is used elsewhere, the behavior might be a bit odd. What do you think about passing `None` as the second argument to `lookupTable->lookup` on L4074 instead?

File: test/Interop/Cxx/namespace/class-inline-namespace-irgen.swift
Comment (zoecarver) [inline test/Interop/Cxx/namespace/class-inline-namespace-irgen.swift:6]: I love this new way to write tests :)

File: test/Interop/Cxx/namespace/class-inline-namespace-irgen.swift
Comment (zoecarver) [inline test/Interop/Cxx/namespace/class-inline-namespace-irgen.swift:8]: `string.h` is a header that may already be in our search paths, so maybe another name is better.

File: test/Interop/Cxx/namespace/class-inline-namespace-irgen.swift
Comment (zoecarver) [inline test/Interop/Cxx/namespace/class-inline-namespace-irgen.swift:24]: This doesn't seem like a really necessary part of the test. It's sort of complicated. It might be better to just use simple types/functions.

File: test/Interop/Cxx/namespace/class-inline-namespace-irgen.swift
Comment (zoecarver) [inline test/Interop/Cxx/namespace/class-inline-namespace-irgen.swift:26]: Generally we use the `TitleCase` format for types. I know this is a test... but might as well start off right.

File: test/Interop/Cxx/namespace/class-inline-namespace-irgen.swift
Comment (zoecarver) [inline test/Interop/Cxx/namespace/class-inline-namespace-irgen.swift:38]: Can you mark these as the end of the namespaces they close?

**Review by zoecarver — COMMENTED** [review]
Some suggestions for tests to add:
- [x] Extending an inline namespace in Swift.
- [x] A type with the same name in two inline namespaces (this might be an ODR violation...).
- [x] A function inside of an inline namespace. 
- [x] A test for `std.inner` and a test for `std.__1.__2.inner`.
- [x] If it's allowed, an inline namespace that is at the top level. 
- [x] (Optional) a namespace inside of an inline namespace inside of a namespace.

File: test/Interop/Cxx/namespace/class-inline-namespace-irgen.swift
Comment (zoecarver) [inline test/Interop/Cxx/namespace/class-inline-namespace-irgen.swift:12]: Also, let's not name this `std`. 

A more general comment: can you update all the names in this test to describe why they're interesting? For example, something like this:
```
namespace Parent {
struct InParent {};
namespace InlineChild {
struct InInlineChild {};
namespace SecondInlineChild {
struct InSecondInlineChild {};
} // namespace SecondInlineChild
} // namespace InlineChild
} // namespace Parent
```

File: test/Interop/Cxx/namespace/inline-namespace-ambiguity-error.swift
Comment (zoecarver) [inline test/Interop/Cxx/namespace/inline-namespace-ambiguity-error.swift:3]: (Assuming I'm reading this right...) I suppose it's OK to keep this test, but to be honest, I'm not sure it provides a lot of value. In this test suite we don't want to verify that clang correctly diagnoses errors. We could have a test that we correctly propagate clang diagnostics in general, but we don't need to verify each diagnostic. That is really the job of the clang test suite.

Comment (hyp) [reply test/Interop/Cxx/namespace/inline-namespace-ambiguity-error.swift:3]: That's not a clang error, that's a swift error.

Comment (zoecarver) [reply test/Interop/Cxx/namespace/inline-namespace-ambiguity-error.swift:3]: Ah, OK. In that case ignore this comment :)

File: lib/ClangImporter/ClangImporter.cpp
Comment (hyp) [reply lib/ClangImporter/ClangImporter.cpp:4278]: My original implementation did that, so I can just go back to that.

Comment (zoecarver) [reply lib/ClangImporter/ClangImporter.cpp:4278]: Perfect. I think this might actually be faster too, because this way we don't filter twice.

Comment (zoecarver) [reply lib/ClangImporter/ClangImporter.cpp:4278]: (But I suppose we have a bigger vector, so maybe not...)

**Comment by hyp** [comment]
@swift-ci please test

File: lib/ClangImporter/ClangImporter.cpp
Comment (zoecarver) [inline lib/ClangImporter/ClangImporter.cpp:4076]: Did you run into problems with this? Macros can't be members so this should be fine.

File: lib/AST/NameLookup.cpp
Comment (zoecarver) [inline lib/AST/NameLookup.cpp:1464]: I'm surprised this works at all.. we need to combine `allFound` with whatever we get from `Table.find()` below.

File: lib/ClangImporter/ClangImporter.cpp
Comment (hyp) [reply lib/ClangImporter/ClangImporter.cpp:4076]: yes, this broke the stdlib build.

File: test/Interop/Cxx/namespace/class-inline-namespace-irgen.swift
Comment (zoecarver) [inline test/Interop/Cxx/namespace/class-inline-namespace-irgen.swift:91]: Unfortunately, these checks aren't going to be super portable. On some platforms this will be an `i32`. You can either use regex or maybe convert to an execution test?

File: test/Interop/Cxx/namespace/class-inline-namespace-irgen.swift
Comment (zoecarver) [inline test/Interop/Cxx/namespace/class-inline-namespace-irgen.swift:89]: Same here: on Windows this is going to have a different name/abi :/

**Review by zoecarver — COMMENTED** [review]
Thank you for overhauling the test. They look great now!

File: lib/ClangImporter/ClangImporter.cpp
Comment (zoecarver) [reply lib/ClangImporter/ClangImporter.cpp:4076]: Hmm, that's super interesting. Do you know how it broke the build or what decl it was? I might go investigate after this lands..

File: lib/AST/NameLookup.cpp
Comment (hyp) [reply lib/AST/NameLookup.cpp:1464]: what kind of lookup would that scenario cover?

Comment (hyp) [reply lib/AST/NameLookup.cpp:1464]: maybe an ambiguous lookup between extension in Swift and the original C++ namespace? I can try that

File: lib/ClangImporter/ClangImporter.cpp
Comment (hyp) [reply lib/ClangImporter/ClangImporter.cpp:4076]: it was `p_starttime`

Comment (hyp) [reply lib/ClangImporter/ClangImporter.cpp:4076]: which I believe is a macro

File: test/Interop/Cxx/namespace/class-inline-namespace-irgen.swift
Comment (hyp) [reply test/Interop/Cxx/namespace/class-inline-namespace-irgen.swift:89]: yes, good point, the tests are failing because of that, so I'll fix it

File: lib/AST/NameLookup.cpp
Comment (zoecarver) [reply lib/AST/NameLookup.cpp:1464]: Sorry, I did kind of a rushed review. I should have explained more carefully. Here's the senerio I'm thinking of: 

```C++
namespace A {
inline namespace B {
void test(int, int);
}
}
```
```swift
extension A {
  func test(_ _: CInt) { }
}

func test() { A.test(0) } // not found
```

File: lib/ClangImporter/ClangImporter.cpp
Comment (zoecarver) [reply lib/ClangImporter/ClangImporter.cpp:4076]: That's super interesting... so it's a macro defined to a member of a struct. I guess this is the correct way to handle this...

Comment (zoecarver) [reply lib/ClangImporter/ClangImporter.cpp:4076]: I wish this could just be a `ranges::filter(isa-named-decl) | ranges::filter(...)`

File: lib/AST/NameLookup.cpp
Comment (zoecarver) [reply lib/AST/NameLookup.cpp:1464]: Actually, I realized, you don't even need an inline namespace, I think. You could just get this to break by doing the following:
```
namespace A {
void test(int, int);
}

extension A {
  func test(_ _: CInt) { }
}

func test() { A.test(0) } // not found
```

**Comment by hyp** [comment]
@swift-ci please test

**Comment by hyp** [comment]
@swift-ci please test

File: lib/AST/NameLookup.cpp
Comment (zoecarver) [inline lib/AST/NameLookup.cpp:1467]: If you're going to duplicate the code anyway, might as well unconditionally take this path.

**Review by zoecarver — APPROVED** [review]
Looks good. Thanks for the diligent updates, Alex!

File: lib/AST/NameLookup.cpp
Comment (hyp) [reply lib/AST/NameLookup.cpp:1467]: The code isn't duplicated, the regular path does something additional - it marks the name as `markLazilyComplete` in the table.

Comment (zoecarver) [reply lib/AST/NameLookup.cpp:1467]: OK, so, if we know that there are no lazy members (i.e., all members are in the lookup table for a given name), then we don't have to emit a namespace lookup request? Makes sense.

**Comment by hyp** [comment]
@swift-ci please test

**Comment by hyp** [comment]
@swift-ci please test

**Comment by hyp** [comment]
@swift-ci please test source compatibility

**Comment by zoecarver** [comment]
@hyp I don't think a source compatibility check is really necessary here. It seems very unlikely this will break source compatibility as we have few projects that use Objective-C interop and none that use C++ interop. 

I think the failures you're seeing for source compatibility are failing on main and will be resolved by #40143. Anyway, I think you're good to merge if you want.

**Comment by hyp** [comment]
> @hyp I don't think a source compatibility check is really necessary here. It seems very unlikely this will break source compatibility as we have few projects that use Objective-C interop and none that use C++ interop.
> 
> I think the failures you're seeing for source compatibility are failing on main and will be resolved by #40143. Anyway, I think you're good to merge if you want.

I know, I saw that the same failures are in the baseline :) I was just curious to see what the source compatibility test would do.

**Comment by zoecarver** [comment]
I know I just approved this patch, but I also just had a thought: this will crash if someone tries to call a function in an inline namespace from the parent namespace, right? If so, that's really not a great failure mode on our part. Ideally we'd have a diagnostic, but it probably doesn't make sense to spend time adding that, as we want to fix this issue pretty soon. How hard would it be to filter out `FuncDecl`s when filter on the inline parent? Maybe one of us can make a patch to do this next week (or maybe it's not needed depending on how quickly we plan to fix the underlying issue).

**Comment by hyp** [comment]
> If so, that's really not a great failure mode on our part. Ideally we'd have a diagnostic, but it probably doesn't make sense to spen

It produces a diagnostic , unless I'm misunderstanding your scenario. One of the test cases covers it.

**Comment by zoecarver** [comment]
> It produces a diagnostic , unless I'm misunderstanding your scenario. One of the test cases covers it.

Ah, no, you're right. It doesn't crash. I was mis remembering.

**Pull request merged**
